### PR TITLE
Add SELinux and systemd checks for pulp3 filesystem

### PR DESCRIPTION
**Pull request by swadeley**
Hello

Some low level checks to ensure the new filesystem for Pulp3 is ready for the migration from Pulp2

### CODE REVIEW

**Comment by swadeley** [comment]
robottelo]$ pytest   --show-capture=no tests/foreman/sys/test_pulp3_filesystem.py 
============================ test session starts =======================================================
platform linux -- Python 3.9.2, pytest-6.2.2, py-1.10.0, pluggy-0.13.1
shared_function enabled - OFF - scope:  - storage: file
rootdir: /robot-test/robottelo, configfile: pyproject.toml
plugins: xdist-2.2.0, forked-1.3.0, services-2.2.1, cov-2.11.1, ibutsu-1.13, mock-3.5.1, reportportal-5.0.7
collected 1 item                                                                                                                                                                                  

tests/foreman/sys/test_pulp3_filesystem.py .                                                                                                                                                [100%]
===================================== 1 passed in 12.97s ====================================

**Review by omkarkhatavkar — APPROVED** [review]
ACK

**Review by ekohl — COMMENTED** [review]
I don't know what this actually tries to assert. It seems odd to me to combine SELinux and pulpcore checks.

I would say that you would write a check that verifies SELinux is enforcing and that there are no logged denials (AVCs). Then in another check you verify Pulp 3 is correctly installed.

As for checks, you can also check if `/var/lib/pulp` is correctly created. https://github.com/theforeman/puppet-pulpcore#installation-layout does describe the layout (also shipped in `/usr/share/foreman-installer/modules/pulpcore/README.md`). Just below that are the services that should be present.

**Comment by swadeley** [comment]
Thank you @ekohl  for the feedback

File: tests/foreman/sys/test_pulp3_filesystem.py
Comment (ekohl) [inline tests/foreman/sys/test_pulp3_filesystem.py:29]:
```suggestion
    :expectedresults: SELinux is enabled and there are no denials
```

File: tests/foreman/sys/test_pulp3_filesystem.py
Comment (ekohl) [inline tests/foreman/sys/test_pulp3_filesystem.py:47]: I think this needs an update

File: tests/foreman/sys/test_pulp3_filesystem.py
Comment (ekohl) [inline tests/foreman/sys/test_pulp3_filesystem.py:71]: I think this is hard to debug when it fails. This is probably easier:
```suggestion
        directories = ['/var/lib/pulp', '/...']
        for directory in directories:
            assert connection.run('test -d ' + directory).return_code == 0
```

That gives you a context variable `directory`. I think you can even write:
```python
assert connection.run('test -d ' + directory).return_code == 0, directory + " must exist"
```

Also, looking at this code I'd consider using [pytest-testinfra](https://github.com/pytest-dev/pytest-testinfra) in the future.

Comment (swadeley) [reply tests/foreman/sys/test_pulp3_filesystem.py:71]: Hello @ekohl

Thank you for feedback, will test that idea.

BTW, while testing I discovered there is no `/var/lib/pulp/pulpcore_static` on my test system:

`~]# ls /var/lib/pulp/
0005_puppet_module_name_change.txt  content  importers  katello-export  packages   static        tmp     uploads
assets                              exports  imports    media           published  sync_imports  uploas`

File: tests/foreman/sys/test_pulp3_filesystem.py
Comment (swadeley) [reply tests/foreman/sys/test_pulp3_filesystem.py]: Fixed , thank you

File: tests/foreman/sys/test_pulp3_filesystem.py
Comment (swadeley) [reply tests/foreman/sys/test_pulp3_filesystem.py]: yep, fixed, thanks

File: tests/foreman/sys/test_pulp3_filesystem.py
Comment (ekohl) [inline tests/foreman/sys/test_pulp3_filesystem.py:72]: Btw, normally the pattern is to use small and very focused tests. https://docs.pytest.org/en/stable/parametrize.html#parametrize-basics describes it. So you would introduce a fixture `connection` and then the method becomes:
```python
@pytest.mark.parametrize('directory', [....])
def test_pulp_directory_exists(connection, directory):
    assert connection.run('test -d ' + directory).return_code == 0
```

However, I don't think robotello follows this pattern.

File: tests/foreman/sys/test_pulp3_filesystem.py
Comment (JacobCallahan) [inline tests/foreman/sys/test_pulp3_filesystem.py:9]: You'll need to add the metadata noting who owns this file

File: tests/foreman/sys/test_pulp3_filesystem.py
Comment (JacobCallahan) [reply tests/foreman/sys/test_pulp3_filesystem.py:72]: Robotello does encourage parameterization. However, for this test, the amount of work done before this point would make parameterization a less than ideal choice.

File: tests/foreman/sys/test_pulp3_filesystem.py
Comment (JacobCallahan) [inline tests/foreman/sys/test_pulp3_filesystem.py:74]: Can you use f-strings for the above two lines?

**Review by JacobCallahan — CHANGES_REQUESTED** [review]

File: tests/foreman/sys/test_pulp3_filesystem.py
Comment (swadeley) [reply tests/foreman/sys/test_pulp3_filesystem.py:9]: OK, done

File: tests/foreman/sys/test_pulp3_filesystem.py
Comment (swadeley) [reply tests/foreman/sys/test_pulp3_filesystem.py]: OK, done

File: tests/foreman/sys/test_pulp3_filesystem.py
Comment (swadeley) [reply tests/foreman/sys/test_pulp3_filesystem.py]: I did that but now changed to use f strings

File: tests/foreman/sys/test_pulp3_filesystem.py
Comment (ekohl) [inline tests/foreman/sys/test_pulp3_filesystem.py:77]:
```suggestion
            ), f'{directory} must exist.'
```

Comment (swadeley) [reply tests/foreman/sys/test_pulp3_filesystem.py:77]: oh dear

Comment (swadeley) [reply tests/foreman/sys/test_pulp3_filesystem.py]: fixed, thank you

**Comment by swadeley** [comment]
robottelo]$ pytest --show-capture=no tests/foreman/sys/test_pulp3_filesystem.py 
================================ test session starts =================================================
platform linux -- Python 3.9.2, pytest-6.2.2, py-1.10.0, pluggy-0.13.1
shared_function enabled - OFF - scope:  - storage: file
rootdir: /robot-test/robottelo, configfile: pyproject.toml
plugins: forked-1.3.0, services-2.2.1, cov-2.11.1, mock-3.5.1, ibutsu-1.14.1, reportportal-5.0.8, xdist-2.2.1
collected 2 items                                                                                                                                                                                 

tests/foreman/sys/test_pulp3_filesystem.py ..                                                                                                                                               [100%]

========================================= 2 passed in 22.76s ====================================

**Comment by swadeley** [comment]
trigger: test-robottelo
pytest -n2: tests/foreman/sys/test_pulp3_filesystem.py

File: tests/foreman/sys/test_pulp3_filesystem.py
Comment (ekohl) [reply tests/foreman/sys/test_pulp3_filesystem.py]: We may have updated that later on. I think an update of Pulpcore is worked on for 6.9 but not 100% sure.

File: tests/foreman/sys/test_pulp3_filesystem.py
Comment (swadeley) [reply tests/foreman/sys/test_pulp3_filesystem.py]: OK, done that. Thank you

**Comment by swadeley** [comment]
trigger: test-robottelo
pytest -n2: tests/foreman/sys/test_pulp3_filesystem.py

File: tests/foreman/sys/test_pulp3_filesystem.py
Comment (ekohl) [inline tests/foreman/sys/test_pulp3_filesystem.py:64]: My suggestion was to replace this with:
```python
@pytest.fixture
def connection(timeout=100):
    with get_connection(timeout=timeout) as connection:
        yield connection


@pytest.mark.upgrade 
@pytest.mark.parametrize('directory', [....])
def test_pulp_directory_exists(connection, directory):
    """Check Pulp3 directories are present.
    :id: 96a64932-9e89-4063-9d5b-55c811375361
    :expectedresults: Directories are present
    """
    assert connection.run(f'test -d {directory}').return_code == 0
```

**Comment by swadeley** [comment]
robottelo]$ pytest --show-capture=no tests/foreman/sys/test_pulp3_filesystem.py 
===================================== test session starts ============================================
platform linux -- Python 3.9.2, pytest-6.2.2, py-1.10.0, pluggy-0.13.1
shared_function enabled - OFF - scope:  - storage: file
rootdir: robot-test/robottelo, configfile: pyproject.toml
plugins: forked-1.3.0, services-2.2.1, cov-2.11.1, mock-3.5.1, ibutsu-1.14.1, reportportal-5.0.8, xdist-2.2.1
collected 3 items                                                                                                                                                                                 

tests/foreman/sys/test_pulp3_filesystem.py ...                                                                                                                                              [100%]

============================================== 3 passed in 24.15s =============================

**Comment by ekohl** [comment]
> robottelo]$ pytest --show-capture=no tests/foreman/sys/test_pulp3_filesystem.py

I'd suggest using `-v` to show individual tests as well when you use parametrized tests.

**Review by JacobCallahan — APPROVED** [review]
ACK. Merge pending @ekohl's suggested parameterization change

**Comment by swadeley** [comment]
robottelo]$ pytest --show-capture=no -v tests/foreman/sys/test_pulp3_filesystem.py 
================================== test session starts ==========================================
platform linux -- Python 3.9.2, pytest-6.2.2, py-1.10.0, pluggy-0.13.1
cachedir: .pytest_cache
shared_function enabled - OFF - scope:  - storage: file
rootdir: robot-test/robottelo, configfile: pyproject.toml
plugins: forked-1.3.0, services-2.2.1, cov-2.11.1, mock-3.5.1, ibutsu-1.14.1, reportportal-5.0.8, xdist-2.2.1
collected 3 items                                                                                                                                                                                 

tests/foreman/sys/test_pulp3_filesystem.py::test_selinux_status PASSED                                                                                                                      [ 33%]
tests/foreman/sys/test_pulp3_filesystem.py::test_pulp_directory_exists PASSED                                                                                                               [ 66%]
tests/foreman/sys/test_pulp3_filesystem.py::test_pulp_service_definitions PASSED                                                                                                            [100%]

==================================== 3 passed in 24.74s ===========================================

File: tests/foreman/sys/test_pulp3_filesystem.py
Comment (swadeley) [reply tests/foreman/sys/test_pulp3_filesystem.py]: OK, done

Comment (ekohl) [reply tests/foreman/sys/test_pulp3_filesystem.py]: I still don't see `@pytest.mark.parametrize('directory', [....])` and I've given the exact code that I think should work twice. Does it not work?

Comment (swadeley) [reply tests/foreman/sys/test_pulp3_filesystem.py]: Hi @ekohl, as @JacobCallahan is a tier two reviewer for this project I will defer to him. See https://github.com/SatelliteQE/robottelo/pull/8400#discussion_r598063542

Thank you

Comment (ekohl) [reply tests/foreman/sys/test_pulp3_filesystem.py]: And I addressed that by suggesting to split off all the additional work. In https://github.com/SatelliteQE/robottelo/pull/8400#pullrequestreview-618941852 @JacobCallahan was very explicit that you should implement my suggestion. That's why I'm confused.

Comment (JacobCallahan) [reply tests/foreman/sys/test_pulp3_filesystem.py]: Correct, the directory test should be parameterized with the values you currently have in the directories list. This will remove the need for a loop within the test itself.

Comment (swadeley) [reply tests/foreman/sys/test_pulp3_filesystem.py]: >  @JacobCallahan was very explicit that you should implement my suggestion. That's why I'm confused.

"the amount of work done before this point would make parameterization **a less than ideal choice**."   that made me think the opposite. Anyway, I have done it now.

**Comment by swadeley** [comment]
robottelo]$ pytest --show-capture=no -v tests/foreman/sys/test_pulp3_filesystem.py 
=============================== test session starts ============================================
platform linux -- Python 3.9.2, pytest-6.2.2, py-1.10.0, pluggy-0.13.1
cachedir: .pytest_cache
shared_function enabled - OFF - scope:  - storage: file
rootdir: robot-test/robottelo, configfile: pyproject.toml
plugins: forked-1.3.0, services-2.2.1, cov-2.11.1, mock-3.5.1, ibutsu-1.14.1, reportportal-5.0.8, xdist-2.2.1
collected 7 items                                                                                                                                                                                 

tests/foreman/sys/test_pulp3_filesystem.py::test_selinux_status PASSED                                                                                                                      [ 14%]
tests/foreman/sys/test_pulp3_filesystem.py::test_pulp_directory_exists[/var/lib/pulp] PASSED                                                                                                [ 28%]
tests/foreman/sys/test_pulp3_filesystem.py::test_pulp_directory_exists[/var/lib/pulp/assets] PASSED                                                                                         [ 42%]
tests/foreman/sys/test_pulp3_filesystem.py::test_pulp_directory_exists[/var/lib/pulp/media] PASSED                                                                                          [ 57%]
tests/foreman/sys/test_pulp3_filesystem.py::test_pulp_directory_exists[/var/lib/pulp/tmp] PASSED                                                                                            [ 71%]
tests/foreman/sys/test_pulp3_filesystem.py::test_pulp_directory_exists[/var/lib/pulp/static] PASSED                                                                                         [ 85%]
tests/foreman/sys/test_pulp3_filesystem.py::test_pulp_service_definitions PASSED                                                                                                            [100%]

=========================================== 7 passed in 32.25s ==================================

File: tests/foreman/sys/test_pulp3_filesystem.py
Comment (JacobCallahan) [reply tests/foreman/sys/test_pulp3_filesystem.py]: That is the amount of work in that test, meaning the number of actions taken before it iterates over directories.

File: tests/foreman/sys/test_pulp3_filesystem.py
Comment (ekohl) [inline tests/foreman/sys/test_pulp3_filesystem.py:25]: > =========================================== 7 passed in 32.25s ==================================

Sadly it does come with a price. This very likely comes from `get_connection` which is now called more often. You can consider making the `connection` fixture a session fixture if it is reusable. I think [pytest-xdist](https://pypi.org/project/pytest-xdist) uses multiple processes so thread safety isn't a huge concern. If [pytest-parallel](https://pypi.org/project/pytest-parallel) is used, it may be. The documentation says it uses a separate channel for every command: https://docs.paramiko.org/en/stable/api/client.html#paramiko.client.SSHClient.exec_command. That means it's probably safe.

I'd try this to see what kind of performance impact it has:

```suggestion
@pytest.fixture(scope="session")
```

**Review by ekohl — APPROVED** [review]
> tests/foreman/sys/test_pulp3_filesystem.py::test_pulp_directory_exists[/var/lib/pulp] PASSED [ 28%]
> tests/foreman/sys/test_pulp3_filesystem.py::test_pulp_directory_exists[/var/lib/pulp/assets] PASSED [ 42%]
> tests/foreman/sys/test_pulp3_filesystem.py::test_pulp_directory_exists[/var/lib/pulp/media] PASSED [ 57%]
> tests/foreman/sys/test_pulp3_filesystem.py::test_pulp_directory_exists[/var/lib/pulp/tmp] PASSED [ 71%]
> tests/foreman/sys/test_pulp3_filesystem.py::test_pulp_directory_exists[/var/lib/pulp/static] PASSED [ 85%]

Here you can clearly see why it's a benefit: it's very easy to figure out which one failed (if it did).

File: tests/foreman/sys/test_pulp3_filesystem.py
Comment (swadeley) [reply tests/foreman/sys/test_pulp3_filesystem.py]: OK, it is faster: = 7 passed in 21.22s =

**Comment by swadeley** [comment]
robottelo]$ pytest --show-capture=no -v tests/foreman/sys/test_pulp3_filesystem.py 
================================ test session starts ==================================
platform linux -- Python 3.9.2, pytest-6.2.2, py-1.10.0, pluggy-0.13.1 
cachedir: .pytest_cache
shared_function enabled - OFF - scope:  - storage: file
rootdir: robot-test/robottelo, configfile: pyproject.toml
plugins: forked-1.3.0, services-2.2.1, cov-2.11.1, mock-3.5.1, ibutsu-1.14.1, reportportal-5.0.8, xdist-2.2.1
collected 7 items                                                                                                                                                                                 

tests/foreman/sys/test_pulp3_filesystem.py::test_selinux_status PASSED                                                                                                                      [ 14%]
tests/foreman/sys/test_pulp3_filesystem.py::test_pulp_directory_exists[/var/lib/pulp] PASSED                                                                                                [ 28%]
tests/foreman/sys/test_pulp3_filesystem.py::test_pulp_directory_exists[/var/lib/pulp/assets] PASSED                                                                                         [ 42%]
tests/foreman/sys/test_pulp3_filesystem.py::test_pulp_directory_exists[/var/lib/pulp/media] PASSED                                                                                          [ 57%]
tests/foreman/sys/test_pulp3_filesystem.py::test_pulp_directory_exists[/var/lib/pulp/tmp] PASSED                                                                                            [ 71%]
tests/foreman/sys/test_pulp3_filesystem.py::test_pulp_directory_exists[/var/lib/pulp/static] PASSED                                                                                         [ 85%]
tests/foreman/sys/test_pulp3_filesystem.py::test_pulp_service_definitions PASSED                                                                                                            [100%]

================================== 7 passed in 21.22s ======================================

**Review by ekohl — APPROVED** [review]
I'll leave it up to the robottelo maintainers to answer this, but connection can be a useful fixture in many places and you can consider moving it to `conftest.py`. Can also be done as part of a follow up PR of course.

**Comment by JacobCallahan** [comment]
> I'll leave it up to the robottelo maintainers to answer this, but connection can be a useful fixture in many places and you can consider moving it to `conftest.py`. Can also be done as part of a follow up PR of course.

That's a valid point, especially considering that the fixture is session scoped within the module.

**Review by mshriver — APPROVED** [review]

**Comment by mshriver** [comment]
Thanks @ekohl for all of your suggestions and review here, we really appreciate that!  I like the discussion here happening about the use of this connection fixture, and the pattern it provides for this test.  Moving the fixture into one of our `pytest_fixtures/` modules gets outside the scope of what @swadeley is trying to contribute for this test coverage, so I'm going to merge the PR as-is.

Our plans for system under test treatment within robottelo are to move toward a specific instance interaction, instead of the ssh 'helper' methods like `get_connection`.  The Satellite/Capsule instances that are the target of the test functions will be provided through a fixture, and ssh/api/cli/db/UI interactions will go through that instance. 

I'm not sure an interim move to using a session scoped ssh connection is worthwhile, but it might be a fairly straightforward swap for some appreciable gains in execution time for ssh heavy tests.

**Comment by swadeley** [comment]
Thanks all

**Comment by ekohl** [comment]
Thanks for that insight @mshriver! For reference, in our upstream testing we're developing [smoker](https://github.com/theforeman/smoker) as a tool and is used in our CI pipelines. Since both are pytest based, it's interesting to think about these problems. Smoker's goal is to be much more functional testing and our current goal is to add client registration tests (using podmain containers). At some point we also want to port over our [content syncing tests](https://github.com/theforeman/forklift/blob/master/bats/fb-katello-content.bats). If we do end up using SSH-based tests on hosts, I'm fairly certain we'll use [pytest-testinfra](https://github.com/pytest-dev/pytest-testinfra).

Back when I still was in the Brno office I have tried to work with QE to align tooling but the gap was large and time was short so it hasn't happened yet. However, I still think that more automated testing in gating pipelines is a worthy goal. Well, more than that. IMHO it's a hard requirement in 2021 to develop and release on time.

I introduced Smoker because it was small and (relatively) easy to integrate in our pipelines. It uses pytest because it's great, but also because it's a technology that QE knows and uses. That means a future merge should be possible and part of me still wants a single tool. There will be challenges so building a small tool that allowed us to iterate was a pragmatic choice.

File: tests/foreman/sys/test_pulp3_filesystem.py
Comment (swadeley) [reply tests/foreman/sys/test_pulp3_filesystem.py]: https://github.com/SatelliteQE/robottelo/pull/8682

**Pull request merged**
